File: internal/authx/awsquery.go
package authx

import (
	"bytes"
	"context"
	"encoding/json"
	"encoding/xml"
	"errors"
	"fmt"
	"net/http"
	"net/url"
	"sort"
	"strconv"
	"strings"

	"github.com/roee-hersh/hallpass/internal/httpx"
	"github.com/roee-hersh/hallpass/internal/integration"
)

// AWSError is a decoded AWS service error (XML for Query-protocol services
// such as IAM and STS, JSON for JSON-1.1 services such as Identity Store).
type AWSError struct {
	Status  int
	Code    string
	Message string
	// RetryAfterSeconds is set by services that send it (Identity Center).
	RetryAfterSeconds int
}

func (e *AWSError) Error() string {
	s := fmt.Sprintf("aws: HTTP %d %s", e.Status, e.Code)
	if e.Message != "" {
		s += ": " + truncate(e.Message, 200)
	}
	return s
}

// Throttled reports whether the error is a rate limit.
func (e *AWSError) Throttled() bool {
	switch e.Code {
	case "Throttling", "ThrottlingException", "RequestLimitExceeded", "TooManyRequestsException", "RequestThrottled", "RequestThrottledException":
		return true
	}
	return e.Status == 429
}

// AccessDenied reports whether the caller lacks permission.
func (e *AWSError) AccessDenied() bool {
	switch e.Code {
	case "AccessDenied", "AccessDeniedException", "UnauthorizedAccess", "UnrecognizedClientException", "InvalidClientTokenId", "ExpiredToken", "ExpiredTokenException", "SignatureDoesNotMatch", "InvalidSignatureException", "IncompleteSignature", "AuthFailure":
		return true
	}
	return e.Status == 401 || e.Status == 403
}

type xmlErrorResponse struct {
	Error struct {
		Code    string `xml:"Code"`
		Message string `xml:"Message"`
	} `xml:"Error"`
	// Some services put the error at the top level.
	Code    string `xml:"Code"`
	Message string `xml:"Message"`
}

// DecodeXMLError parses a Query-protocol error body.
func DecodeXMLError(status int, body []byte) *AWSError {
	var x xmlErrorResponse
	_ = xml.Unmarshal(body, &x)
	code, msg := x.Error.Code, x.Error.Message
	if code == "" {
		code, msg = x.Code, x.Message
	}
	if code == "" {
		code = "HTTP" + strconv.Itoa(status)
	}
	return &AWSError{Status: status, Code: code, Message: msg}
}

// DecodeJSONError parses a JSON-1.1 error body and the x-amzn-ErrorType
// header.
func DecodeJSONError(status int, header http.Header, body []byte) *AWSError {
	var j struct {
		Type              string `json:"__type"`
		Message           string `json:"message"`
		MessageCap        string `json:"Message"`
		RetryAfterSeconds int    `json:"RetryAfterSeconds"`
	}
	_ = json.Unmarshal(body, &j)
	code := j.Type
	if h := header.Get("x-amzn-ErrorType"); h != "" {
		code = h
	}
	// "com.amazonaws.service#ThrottlingException:http://..." -> ThrottlingException
	if i := strings.Index(code, "#"); i >= 0 {
		code = code[i+1:]
	}
	if i := strings.Index(code, ":"); i >= 0 {
		code = code[:i]
	}
	if code == "" {
		code = "HTTP" + strconv.Itoa(status)
	}
	msg := j.Message
	if msg == "" {
		msg = j.MessageCap
	}
	return &AWSError{Status: status, Code: code, Message: msg, RetryAfterSeconds: j.RetryAfterSeconds}
}

// ClassifyAWSError maps an AWS call failure to an integration error.
func ClassifyAWSError(err error) *integration.Error {
	var ae *AWSError
	if errors.As(err, &ae) {
		switch {
		case ae.Throttled():
			return integration.Wrap(integration.CodeUpstreamRateLimit, err, "AWS throttled the request (%s)", ae.Code)
		case ae.AccessDenied():
			return integration.Wrap(integration.CodeCredentialRejected, err, "AWS rejected hallpass's credential or denied the call (%s)", ae.Code)
		case ae.Status >= 500:
			return integration.Wrap(integration.CodeUpstreamError, err, "AWS returned %s", ae.Code)
		default:
			return integration.Wrap(integration.CodeUpstreamError, err, "AWS returned %s", ae.Code)
		}
	}
	return httpx.Classify(err)
}

// QueryParams builds Query-protocol form values. Lists are encoded as
// Name.member.N (1-based); nested lists of structs are expressed by the
// caller as flattened keys such as "ContextEntries.member.1.ContextKeyName".
type QueryParams map[string]any

// Form encodes the params with Action and Version.
func (p QueryParams) Form(action, version string) url.Values {
	v := url.Values{"Action": {action}, "Version": {version}}
	keys := make([]string, 0, len(p))
	for k := range p {
		keys = append(keys, k)
	}
	sort.Strings(keys)
	for _, k := range keys {
		switch val := p[k].(type) {
		case string:
			v.Set(k, val)
		case int:
			v.Set(k, strconv.Itoa(val))
		case int64:
			v.Set(k, strconv.FormatInt(val, 10))
		case bool:
			v.Set(k, strconv.FormatBool(val))
		case []string:
			for i, s := range val {
				v.Set(k+".member."+strconv.Itoa(i+1), s)
			}
		default:
			v.Set(k, fmt.Sprint(val))
		}
	}
	return v
}

// CredentialProvider yields AWS credentials, refreshing as needed.
type CredentialProvider interface {
	Credentials(ctx context.Context) (AWSCredentials, error)
}

// AWSClient signs and sends requests to one AWS service endpoint.
type AWSClient struct {
	HTTP *httpx.Client
	// Endpoint is the full base URL, e.g. https://iam.amazonaws.com.
	Endpoint string
	// Region is the signing region; Service the signing name.
	Region  string
	Service string
	Creds   CredentialProvider
}

// Query sends a Query-protocol request (form POST) and decodes the XML
// response into out. Errors are *AWSError or transport errors.
func (c *AWSClient) Query(ctx context.Context, action, version string, params QueryParams, out any) error {
	form := params.Form(action, version)
	body := []byte(form.Encode())
	hdr := http.Header{"Content-Type": {"application/x-www-form-urlencoded; charset=utf-8"}, "Accept": {"application/xml"}}
	resp, err := c.send(ctx, body, hdr)
	if err != nil {
		return err
	}
	if resp.Status >= 400 {
		return DecodeXMLError(resp.Status, resp.Body)
	}
	if out != nil {
		if err := xml.Unmarshal(resp.Body, out); err != nil {
			return fmt.Errorf("decode %s response: %w", action, err)
		}
	}
	return nil
}

// JSON11 sends a JSON-1.1 request with X-Amz-Target and decodes the reply.
func (c *AWSClient) JSON11(ctx context.Context, target string, in, out any) error {
	body, err := json.Marshal(in)
	if err != nil {
		return err
	}
	if in == nil {
		body = []byte("{}")
	}
	hdr := http.Header{"Content-Type": {"application/x-amz-json-1.1"}, "X-Amz-Target": {target}}
	resp, err := c.send(ctx, body, hdr)
	if err != nil {
		return err
	}
	if resp.Status >= 400 {
		return DecodeJSONError(resp.Status, resp.Header, resp.Body)
	}
	if out != nil && len(bytes.TrimSpace(resp.Body)) > 0 {
		if err := json.Unmarshal(resp.Body, out); err != nil {
			return fmt.Errorf("decode %s response: %w", target, err)
		}
	}
	return nil
}

func (c *AWSClient) send(ctx context.Context, body []byte, hdr http.Header) (*httpx.Response, error) {
	creds, err := c.Creds.Credentials(ctx)
	if err != nil {
		return nil, err
	}
	idem := true // every call hallpass makes is a read
	req := &httpx.Request{Method: http.MethodPost, Path: c.Endpoint + "/", Body: body, Header: hdr, Idempotent: &idem, Accept4xx: true}
	client := *c.HTTP
	signer := SigV4Signer{Region: c.Region, Service: c.Service}
	client.Auth = func(_ context.Context, r *http.Request) error {
		signer.Sign(r, body, creds)
		return nil
	}
	return client.Do(ctx, req)
}

// DNSSuffix returns the DNS suffix of a partition.
func DNSSuffix(partition string) string {
	switch partition {
	case "aws-cn":
		return "amazonaws.com.cn"
	default:
		return "amazonaws.com"
	}
}

// RegionalEndpoint builds https://{service}.{region}.{suffix}.
func RegionalEndpoint(partition, service, region string) string {
	return "https://" + service + "." + region + "." + DNSSuffix(partition)
}

// IAMEndpoint returns the global IAM endpoint and its signing region.
// The China endpoint is UNVERIFIED.
func IAMEndpoint(partition string) (endpoint, region string) {
	switch partition {
	case "aws-us-gov":
		return "https://iam.us-gov.amazonaws.com", "us-gov-west-1"
	case "aws-cn":
		// UNVERIFIED: China partition IAM endpoint and signing region.
		return "https://iam.cn-north-1.amazonaws.com.cn", "cn-north-1"
	default:
		return "https://iam.amazonaws.com", "us-east-1"
	}
}

File: hallpass-py/src/hallpass/authx/awsquery.py
"""AWS service calls: Query protocol (IAM, STS; form POST, XML replies) and
JSON 1.1 (Identity Store), signed with SigV4, and the errors they return."""

from __future__ import annotations

import copy
from collections.abc import Mapping
from dataclasses import dataclass
from typing import Any, Protocol

from hallpass.authx import xmlutil
from hallpass.authx.sigv4 import AWSCredentials, SigV4Signer
from hallpass.authx.util import INT, STR, go_json_loads, go_json_marshal, go_sprint, go_unmarshal
from hallpass.core.context import Context
from hallpass.core.decision import Code, HallpassError, wrap_error
from hallpass.core.errors import as_error
from hallpass.net import httpx

__all__ = [
    "AWSClient",
    "AWSError",
    "CredentialProvider",
    "QueryParams",
    "classify_aws_error",
    "decode_json_error",
    "decode_xml_error",
    "dns_suffix",
    "iam_endpoint",
    "query_form",
    "regional_endpoint",
]


class AWSError(Exception):
    """A decoded AWS service error."""

    def __init__(self, status: int, code: str, message: str = "", retry_after_seconds: int = 0) -> None:
        self.status = status
        self.code = code
        # Kept but never rendered: for SignatureDoesNotMatch it can carry the
        # canonical request, and error strings end up in logs.
        self.message = message
        self.retry_after_seconds = retry_after_seconds
        super().__init__(str(self))

    def __str__(self) -> str:
        return f"aws: HTTP {self.status} {self.code}"

    def throttled(self) -> bool:
        if self.code in (
            "Throttling",
            "ThrottlingException",
            "RequestLimitExceeded",
            "TooManyRequestsException",
            "RequestThrottled",
            "RequestThrottledException",
        ):
            return True
        return self.status == 429

    def access_denied(self) -> bool:
        if self.code in (
            "AccessDenied",
            "AccessDeniedException",
            "UnauthorizedAccess",
            "UnrecognizedClientException",
            "InvalidClientTokenId",
            "ExpiredToken",
            "ExpiredTokenException",
            "SignatureDoesNotMatch",
            "InvalidSignatureException",
            "IncompleteSignature",
            "AuthFailure",
        ):
            return True
        return self.status in (401, 403)


def decode_xml_error(status: int, body: bytes) -> AWSError:
    """A Query-protocol error body."""
    code = msg = ""
    try:
        # Go ignores the decode error and keeps whatever it decoded.
        root = xmlutil.parse(body, partial=True)
    except xmlutil.XMLError:
        root = None
    if root is not None:
        code, msg = xmlutil.text(root, "Error", "Code"), xmlutil.text(root, "Error", "Message")
        if code == "":
            code, msg = xmlutil.text(root, "Code"), xmlutil.text(root, "Message")
    if code == "":
        code = f"HTTP{status}"
    return AWSError(status, code, msg)


def decode_json_error(status: int, header: httpx.Headers, body: bytes) -> AWSError:
    """A JSON-1.1 error body and the x-amzn-ErrorType header."""
    # Go ignores the decode error and keeps whatever it decoded.
    j, _ = go_unmarshal(body, (("__type", STR), ("message", STR), ("Message", STR), ("RetryAfterSeconds", INT)))
    code = j["__type"]
    h = header.get("x-amzn-ErrorType")
    if h:
        code = h
    # "com.amazonaws.service#ThrottlingException:http://..." -> ThrottlingException
    if "#" in code:
        code = code[code.index("#") + 1 :]
    if ":" in code:
        code = code[: code.index(":")]
    if code == "":
        code = f"HTTP{status}"
    msg = j["message"] or j["Message"]
    return AWSError(status, code, msg, j["RetryAfterSeconds"])


def classify_aws_error(err: BaseException) -> HallpassError:
    ae = as_error(err, AWSError)
    if ae is not None:
        if ae.throttled():
            return wrap_error(Code.UPSTREAM_RATE_LIMIT, err, f"AWS throttled the request ({ae.code})")
        if ae.access_denied():
            return wrap_error(Code.CREDENTIAL_REJECTED, err, f"AWS rejected hallpass's credential or denied the call ({ae.code})")
        return wrap_error(Code.UPSTREAM_ERROR, err, f"AWS returned {ae.code}")
    out = httpx.classify(err)
    assert out is not None
    return out


QueryParams = Mapping[str, Any]


def query_form(params: QueryParams, action: str, version: str) -> dict[str, list[str]]:
    """Query-protocol form values with Action and Version. Lists become
    Name.member.N (1-based); nested structures are flattened by the caller
    ("ContextEntries.member.1.ContextKeyName")."""
    v: dict[str, list[str]] = {"Action": [action], "Version": [version]}
    for k in sorted(params):
        val = params[k]
        if isinstance(val, bool):
            v[k] = ["true" if val else "false"]
        elif isinstance(val, str):
            v[k] = [val]
        elif isinstance(val, int):
            v[k] = [str(val)]
        elif isinstance(val, (list, tuple)) and all(isinstance(s, str) for s in val):
            # Go's []string case.
            for i, s in enumerate(val):
                v[f"{k}.member.{i + 1}"] = [s]
        else:
            v[k] = [go_sprint(val)]
    return v


class CredentialProvider(Protocol):
    def credentials(self, ctx: Context) -> AWSCredentials: ...


@dataclass
class AWSClient:
    """Signs and sends requests to one AWS service endpoint."""

    http: httpx.Client
    # The full base URL, e.g. https://iam.amazonaws.com.
    endpoint: str
    # The signing region and signing name.
    region: str
    service: str
    creds: CredentialProvider

    def query(self, ctx: Context, action: str, version: str, params: QueryParams) -> Any:
        """A Query-protocol request; returns the parsed XML root. Errors are
        AWSError or transport errors."""
        body = httpx.encode_query(query_form(params, action, version)).encode("ascii")
        hdr = {"Content-Type": "application/x-www-form-urlencoded; charset=utf-8", "Accept": "application/xml"}
        resp = self._send(ctx, body, hdr)
        if resp.status >= 400:
            raise decode_xml_error(resp.status, resp.body)
        try:
            return xmlutil.parse(resp.body)
        except xmlutil.XMLError as e:
            raise ValueError(f"decode {action} response: {e}") from None

    def json11(self, ctx: Context, target: str, body_in: Any) -> Any:
        """A JSON-1.1 request with X-Amz-Target; returns the decoded reply
        (None for an empty body)."""
        body = b"{}" if body_in is None else go_json_marshal(body_in)
        hdr = {"Content-Type": "application/x-amz-json-1.1", "X-Amz-Target": target}
        resp = self._send(ctx, body, hdr)
        if resp.status >= 400:
            raise decode_json_error(resp.status, resp.header, resp.body)
        if not resp.body.strip():
            return None
        try:
            return go_json_loads(resp.body)
        except ValueError as e:
            raise ValueError(f"decode {target} response: {e}") from None

    def _send(self, ctx: Context, body: bytes, hdr: Mapping[str, str]) -> httpx.Response:
        creds = self.creds.credentials(ctx)
        client = copy.copy(self.http)
        signer = SigV4Signer(self.region, self.service)

        def auth(_: Context, r: httpx.PreparedRequest) -> None:
            signer.sign(r, body, creds)

        client.auth = auth
        # Every call hallpass makes is a read, so it may be retried.
        return client.do(ctx, httpx.Request(method="POST", path=self.endpoint + "/", body=body, header=dict(hdr), idempotent=True, accept_4xx=True))


def dns_suffix(partition: str) -> str:
    return "amazonaws.com.cn" if partition == "aws-cn" else "amazonaws.com"


def regional_endpoint(partition: str, service: str, region: str) -> str:
    return f"https://{service}.{region}.{dns_suffix(partition)}"


def iam_endpoint(partition: str) -> tuple[str, str]:
    """The global IAM endpoint and its signing region. The China endpoint
    is UNVERIFIED."""
    if partition == "aws-us-gov":
        return "https://iam.us-gov.amazonaws.com", "us-gov-west-1"
    if partition == "aws-cn":
        return "https://iam.cn-north-1.amazonaws.com.cn", "cn-north-1"
    return "https://iam.amazonaws.com", "us-east-1"
